### DISCUSSION TITLE
How to prevent Link component from calling SSR components with `_rsc` at the end of the URL?

### DISCUSSION THREAD

**Discussion by colonder** [vercel/next.js#58357] [Help]
### Summary

I've been using Next's [Link component ](https://nextjs.org/docs/app/api-reference/components/link) all over my code, as per good practices and recommendations. I've also written all of `page.tsx` components to accommodate for SSR e.g. moved client-side components down the tree etc. to maximize performance. Unfortunately, as I noticed in production, and later read in the [docs](https://nextjs.org/docs/app/building-your-application/rendering/server-components), the `<Link/>` component fetches a given page by adding `_rsc=XXXXX` at the end of the href, where `XXX` is some kind of id, regardless of whether it prefetches the page or not. This `_rsc` suffix tells Next.js to fetch a React Server Component optimized binary of the given page that was created during a production build. This behavior confuses Google Search Console greatly, as I have a lot of "duplicate page with the same canonical URL" entries like so:
```
https://www.domain.com/page1_rsc=XXXXX
https://www.domain.com/page2_rsc=XXXXX
```
where in reality their canonical URLs are just `https://www.domain.com/page1` etc. This happens due to Google bot scanning pages, and following links like a normal browser would do. I couldn't find any clue on how to prevent this behavior, so I replaced all `<Link/>` components with a standard `<a>` tag. Is there any way to use `<Link/>` component, retain SSR, and get rid of this `_rsc` suffix? Or tell Google bot that this `_rsc` refers to the same component?

**Comment by Antonio112009** [comment]
I have the same issue. It is annoying that it appears in Google Search Console. 
My Next.js version - 14.0.2-canary8

<img width="113" alt="image" src="https://github.com/vercel/next.js/assets/29264726/d0258cc2-a489-42d5-b743-6475f2cad6b1">

**Reply by Antonio112009** [reply]
Update: adding exclusion from crawling leads to issues in Google Console:
- Blocked by robots.txt
- Server error (5xx)

<img width="782" alt="image" src="https://github.com/vercel/next.js/assets/29264726/ed27a5e0-cd5d-4da5-8f7e-948abf368206">

**Reply by teh23** [reply]
heys folks @colonder @Antonio112009 did u found solution?

**Reply by colonder** [reply]
@teh23 yes, stop using Next.js and switch to some other solution. Jokes aside, no, I did not find anything, however I can see that more and more ppl start complaining about it.

**Reply by colonder** [reply]
@Antonio112009 @teh23 do you have any code sandbox that can reproduce this issue so that a Github issue could be created out of this?

**Reply by Antonio112009** [reply]
Feel free to use the https://nextjs.org 

Before scrolling the page:
<img width="932" alt="image" src="https://github.com/vercel/next.js/assets/29264726/5772c22e-74e8-4352-9914-4fb2ba0cc42f">

After you scrolled to any link:
<img width="1144" alt="image" src="https://github.com/vercel/next.js/assets/29264726/14026f25-559d-4419-ab51-1e8468dcbeb0">

Jokes aside, can somebody create in CodeBox.

**Reply by bennet-org** [reply]
This is becoming a real problem for us as well. Given that it's replicable today on nextjs.com, what specifically is required in the Codebox / for this to be created as an issue?

**Reply by colonder** [reply]
According to this docs page https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-rendering I think if you have any dynamic function on the page, nothing can be done essentially. You'd have to basically use SSG at this point. I'm looking into Astro right now as a replacement for Next.js, at least for landing page.

**Reply by bennet-org** [reply]
But the requests for `_rsc` URLs while scrolling past link elements aren't really the issue here. Googlebot doesn't scroll, so these aren't a problem for SEO. The issue is that the `initialCanonicalUrl` prop in the JSON in the raw HTML source often contains an `_rsc` parameter. Once the page is rendered, the value of the canonical tag i correct, but Google is somehow discovering the `_rsc` versions regardless. We just need the `_rsc` to *never* be included in canonicals or `href` attributes of <a> elements - that's enough for Google to not discover these variants.

**Reply by colonder** [reply]
You might be right, I'm not that proficient with Next.js. If you believe that's the core reason, maybe THAT could be opened as an issue?

**Reply by c0b41** [reply]
@tombennet do you have any example for Canonical Url problem, because i don't see anything like that?

**Reply by bennet-org** [reply]
There's an issue related to `initialCanonicalUrl` here: https://github.com/vercel/next.js/issues/53274

It's not specifically related to the `rsc` param, but I suspect the underlying causes are related (i.e. non-canonical URLs in the `initialCanonicalUrl` prop). I'm still struggling to reliably reproduce the `rsc` appearance in `initialCanonicalUrl` - both Googlebot and my test crawls seems to encounter it only some of the time, which is driving me crazy.

**Reply by c0b41** [reply]
Well that issue more like search engine bot problem

**Comment by albing123** [comment]
Did you find any solution for this? 
@colonder @Antonio112009

**Reply by colonder** [reply]
https://github.com/vercel/next.js/discussions/58357#discussioncomment-8452649

**Reply by david-uriostegui-a** [reply]
> Did you find any solution for this? @colonder @Antonio112009

This should work
`<Link prefetch={true}>`

**Reply by david-uriostegui-a** [reply]
Or use this config
```
// next.config.js
module.exports = {
  experimental: {
    serverComponents: false, 
  },
};

```

**Reply by arfa123** [reply]
> Or use this config
> 
> ```
> // next.config.js
> module.exports = {
>   experimental: {
>     serverComponents: false, 
>   },
> };
> ```

There is no such option available in the Next.js configuration.

**Comment by kstroka** [comment]
Is there any updates on this issue? Did you solve that?

**Comment by icopp** [comment]
In short, you can't, because it's baked into the RSC behavior. There's pending work [to refactor how that works](https://github.com/vercel/next.js/issues/65335#issuecomment-2583401824) from the React team to remove the `?rsc` params/headers from how it works, but that hasn't been released yet.
